lychee-lib/src/client.rs

```diff
@@ -624,13 +636,33 @@ impl Client {
     }
 
     /// Check a `file` URI.
-    pub fn check_file(&self, uri: &Uri) -> Status {
-        if let Ok(path) = uri.url.to_file_path() {
-            if path.exists() {
-                return Status::Ok(StatusCode::OK);
+    pub async fn check_file(&self, uri: &Uri, check_md_frag: Option<FragmentGen>) -> Status {
+        let Ok(path) = uri.url.to_file_path() else {
+            return ErrorKind::InvalidFilePath(uri.clone()).into();
+        };
+        if !path.exists() {
+            return ErrorKind::InvalidFilePath(uri.clone()).into();
+        }
+        let is_md = path
+            .extension()
+            .map_or(false, |ext| ext.eq_ignore_ascii_case("md"));
+
+        if let (true, Some(frag), Some(frag_gen)) = (is_md, uri.url.fragment(), check_md_frag) {
+            let mut content = String::new();
+            File::open(path)
+                .await
+                .expect("Couldn't open a file expected to exist.")
+                .read_to_string(&mut content)
+                .await
+                .expect("Couldn't read a linked file.");
```

Comment:
As you mentioned, this is the key part where we might lose a lot of performance, because we potentially parse the file many times. That's an `O(n^2)` algorithm, which scales badly with the number of input files.

As an alternative, we could introduce a `FragementCache`.
It would be defined as

```rust
type FragmentCache = Arc<Mutex<HashMap<String, HashSet<String>>>>;
```

This way it could be independent of the file extension, so the algorithm becomes

```rust
if let Some(fragement) = uri.url.fragment() {
  self.fragment_cache.contains(fragment)
}
```

The `FragmentCache` would be initialized when the client starts (i.e. in the client constructor and be a member of the client, so that it can be accessed with `self.fragment_cache.contains(fragment)` as seen above.

Reply:
That sounds sensible. I'll add a fragment cache.

I was toying with the idea of finding the fragments in the same pass as links for the files given by the user to be checked, avoiding one having to open and parse these files twice.  I'd guess a common use case is to check many files linking mainly to one another.

It would be nice to pre-populate the cache with the given files in this way, but it would require some larger changes. Do you think it would be worth it, if so how would you recommend going about it?